x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,206 @@
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+
+	"github.com/tharsis/ethermint/crypto/ethsecp256k1"
+
+	evmos "github.com/tharsis/evmos/v2/types"
+)
+
+// OnRecvPacket performs an IBC receive callback.
+func (k Keeper) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	ack exported.Acknowledgement,
+) exported.Acknowledgement {
+	logger := k.Logger(ctx)
+
+	params := k.GetParams(ctx)
+	claimsParams := k.claimsKeeper.GetParams(ctx)
+
+	// check channels from this chain (i.e destination)
+	if !params.EnableWithdraw ||
+		!claimsParams.IsAuthorizedChannel(packet.DestinationChannel) ||
+		claimsParams.IsEVMChannel(packet.DestinationPort) {
+		// return original ACK if:
+		// - withdraw is disabled globally
+		// - channel is not authorized
+		// - channel is an EVM channel
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	// get the recipient account
+	account := k.accountKeeper.GetAccount(ctx, recipient)
+
+	// NOTE: check if the recipient pubkey is a supported key, if it is,
+	// return the original ACK
+
+	// 1.a. no recipient account or no pubkey -> no way to determine the type of address
+	// ==> return success ACK
+	if account != nil &&
+		account.GetPubKey() != nil &&
+		// 1.b: pubkey is eth_secp256k1 (valid ethereum address) or other supported keys
+		// ==> return success ACK
+		(account.GetPubKey().Type() == ethsecp256k1.KeyType ||
+			account.GetPubKey().Type() == (&multisig.LegacyAminoPubKey{}).Type() ||
+			account.GetPubKey().Type() == (&ed25519.PubKey{}).Type()) {
```

Comment:
Do you plan to add more keytypes? Should this be a function (isSupportedKeyType)?